src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -4414,6 +4414,209 @@ class StubGenerator: public StubCodeGenerator {
+    __ srli(t1, R_1, 2);
+    __ shadd(RR_1, t1, t1, t2, 2);
+
+    // U_n is the current checksum
+    const Register U_0 = *++regs, U_1 = *++regs, U_2 = *++regs;
+    pack_26(U_0, U_1, U_2, acc_start, t1, t2);
+
+    static constexpr int BLOCK_LENGTH = 16;
+    Label DONE, LOOP;
+
+    __ mv(t1, checked_cast<u1>(BLOCK_LENGTH));
+    __ blt(length, t1, DONE); {
+      __ bind(LOOP);
+
+      // S_n is to be the sum of U_n and the next block of data
+      const Register S_0 = *++regs, S_1 = *++regs, S_2 = *++regs;
+      __ ld(S_0, Address(input_start, 0));
+      __ ld(S_1, Address(input_start, wordSize));
+
+      __ cad(S_0, S_0, U_0, t1); // Add U_0 to S_0 with carry output to t1
+      __ cadc(S_1, S_1, U_1, t1); // Add U_1 with carry to S_1 with carry output to t1
+      __ add(S_2, U_2, t1);
+
+      __ addi(S_2, S_2, 1);
+
+      const Register U_0HI = *++regs, U_1HI = *++regs;
+
+      // NB: this logic depends on some of the special properties of
+      // Poly1305 keys. In particular, because we know that the top
+      // four bits of R_0 and R_1 are zero, we can add together
+      // partial products without any risk of needing to propagate a
+      // carry out.
+      __ wide_mul(U_0, U_0HI, S_0, R_0);
+      __ wide_madd(U_0, U_0HI, S_1, RR_1, t1, t2);
+      __ wide_madd(U_0, U_0HI, S_2, RR_0, t1, t2);
+
+      __ wide_mul(U_1, U_1HI, S_0, R_1);
+      __ wide_madd(U_1, U_1HI, S_1, R_0, t1, t2);
+      __ wide_madd(U_1, U_1HI, S_2, RR_1, t1, t2);
+
+      __ andi(U_2, R_0, bits2);
+      __ mul(U_2, S_2, U_2);
+
+      // Recycle registers S_0, S_1, S_2
+      regs = (regs.remaining() + S_0 + S_1 + S_2).begin();
+
+      // Partial reduction mod 2**130 - 5
+      __ cad(U_1, U_1, U_0HI, t1); // Add U_0HI to U_1 with carry output to t1
+      __ adc(U_2, U_2, U_1HI, t1);
+      // Sum now in U_2:U_1:U_0.
+      // Dead: U_0HI, U_1HI.
+      regs = (regs.remaining() + U_0HI + U_1HI).begin();
+
+      // U_2:U_1:U_0: += (U_2 >> 2) * 5
+      __ srli(t1, U_2, 2);
+      __ andi(U_2, U_2, bits2); // Clear U_2 except for the first two bits
+      __ shadd(t1, t1, t1, t2, 2); // t1 is impossible to overflow since two leftmost bits are zero'ed in 'srli(t1, U_2, 2)'
+      __ cad(U_0, U_0, t1, t2); // Add t1 (= (U_2 >> 2) * 5) to U_0 with carry output to t2
+      __ cad(U_1, U_1, t2, t2); // Add carry to U_1 with carry output to t2
+      __ add(U_2, U_2, t2);
+
+      __ sub(length, length, checked_cast<u1>(BLOCK_LENGTH));
+      __ addi(input_start, input_start, 2 * wordSize);
+      __ mv(t1, checked_cast<u1>(BLOCK_LENGTH));
+      __ bge(length, t1, LOOP);
+    }
+
+    // Further reduce modulo 2^130 - 5
+    __ srli(t1, U_2, 2);
+    __ shadd(t1, t1, t1, t2, 2); // t1 = U_2 * 5
+    __ cad(U_0, U_0, t1, t2); // U_0 += U_2 * 5 with carry output to t2
+    __ cad(U_1, U_1, t2, t2); // Add carry to U_1 with carry output to t2
+    __ andi(U_2, U_2, bits2);
+    __ add(U_2, U_2, t2); // Add carry to U_2
```

Comment:
What happens if U_2 == 0b11 and t2 == 1? Is it possible?
If it's possible, do we need another (and final) reduce?

Reply:
Maybe there is no chance for this to happen, as in the loop there is already one reducing before quit the loop, the `reduce` here is already the final one?
I'm not quite sure.

Reply:
There are two cases here:
1. No loop cycles were performed: the maximum value in `U_2` before the `reduce` is `0b11`, so it is cleared by `srli`;
2. Some loop cycles were performed: it is possible for `U_2 == 0b11` and `t2 == 1` to happen at the end of loop's last `reduce` block, that's correct. However, at the next reduce outside the loop, `U_2` would have `0b100` value which would be cleared to safe `0b1` by `srli`.

Reply:
Yes, what I'm discussing is the case 2.
Sorry, I'm still bit concerned.

You said `at the next reduce outside the loop, U_2 would have 0b100 value which would be cleared to safe 0b1 by srli.`, but seems the dest register of srli is tmp1 not U_2, am I right?

If I'm right above, then let's continue the discussion.
Let's say at the entry of LINE 4576 `reduce(U_2, U_1, U_0, t1, t2);` (just out of the loop), U_2 == ob11.
Then in reduce(), U_2 will be 0b11 at the entry of LINE 4478 `__ add(U_2, U_2, tmp2);`, and tmp2 == 1. Is it possible?

Reply:
Done.

Reply:
Thanks for updating.
Looks safer, but I'm not sure if it's too conservative or not safe enough.
Can we wait for a while to see how others think about it or what new discoveries have we made about ourselves?

Reply:
I think the caller does a full mod N reduction of your result in `IntegerPolynomial::final Reduce()`.

Reply:
@theRealAph Besides of the reduce discussed above, I still got several other questions about the original implementation in aarch64 (as this implementation is kind of translation from the original one), would you mind help to clarify a bit? Thanks!
1. `__ add(S_2, S_2, 1);` in L7168 at stubGenerator_aarch64.cpp, what's the purpose of this add 1?
2. `wide_mul(U_1, U_1HI, S_0, R_1);  wide_madd(U_1, U_1HI, S_1, R_0);  wide_madd(U_1, U_1HI, S_2, RR_1);` in L7178 at stubGenerator_aarch64.cpp, why S_2*RR_1 does not consider the low 2 bit of R_1, but just the higher bits (which in RR_1)?

Reply:
Seems we don't need the extra reduce introduced in commit `053b7c0`.
Image this situation:
1. in the loop (final round), at the end of reduce `__ add(U_2, U_2, tmp2);`, U_2 is `0b11`, tmp2 is `0b1`, that's the case of maximum value we can get.
2. out of the loop, reduce can never overflow 130 bits anymore, because top 2 bits of 130 bits is `0b00` now.

So I think we are safe to revert before `053b7c0`.

Reply:
About one of the above questions `__ add(S_2, S_2, 1); in L7168 at stubGenerator_aarch64.cpp, what's the purpose of this add 1?`, I think I get it, it's in the paper of https://cr.yp.to/mac/poly1305-20050329.pdf, at section `Conversion and padding`, I missed that part, sorry.

Reply:
> Thanks for the information @theRealAph Seems the finalReduce in java code does not make reduce in intrinsic code here unnecessary, because in the intrinsic code, we only store back 130 bits back to acc_start,

I agree.

Reply:
Thanks for this information. @theRealAph 

@ArsenyBochkarev `We write 131 bits back`, I think it's safe to align riscv implementation with  aarch64 implementation, currently the code is `andi(t2, U_2, bits2);`, this should be changed to `andi(t2, U_2, 0b111);`? And at same time, we are safe to revert before commit 053b7c0.

Reply:
@theRealAph one more question for you (asked above among others), hope it's the last one :)

2. `wide_mul(U_1, U_1HI, S_0, R_1);  wide_madd(U_1, U_1HI, S_1, R_0);  wide_madd(U_1, U_1HI, S_2, RR_1);` in L7178 at stubGenerator_aarch64.cpp, why S_2*RR_1 does not consider the low 2 bit of R_1, but just the higher bits (which in RR_1)?

Reply:
The low 2 bits are 00.

Reply:
Ah, right, all clear! Thanks!

But this makes it harder or not that performant to vectorise poly3015 based on this implementation, as for example R will be R^4 or R^8 (depends on lane numbers we have at runtime), these restrictions designed by the author of poly1305 will not hold anymore.

Reply:
Exactly so. Vectorizing poly1305 is not easy to do at all.

The key paper is Martin Goll and Shay Gueron. _Vectorization of Poly1305 message authentication code_. In 2015 12th International Conference on Information Technology - New Generations, pages 145–150. IEEE, April
2015. 10.1109/ITNG.2015.28.

I have an AArch64 implementation almost ready, but it is complicated.

Reply:
Thanks for sharing!
Can I ask how's the performance improvement?

Reply:
FYI: https://camel-cdr.github.io/rvv-bench-results/canmv_k230/index.html includes data on poly

Reply:
At the moment about 1.8, but I think i can get it up to 2.

Reply:
Tha's great! Thanks!

Reply:
Did the `andi(t2, U_2, 0b111);` and removed conditional reduce.